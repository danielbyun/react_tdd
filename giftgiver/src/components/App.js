import React, { Fragment, useState } from "react";
import { v4 as uuid } from "uuid";

import { Grid, Typography, Button } from "@material-ui/core";
import GiftForm from "./GiftForm";

const App = () => {
  const [currentGifts, setCurrentGifts] = useState([
    {
      id: uuid(),
      name: "PS5",
      recipient: "Mark",
    },
    {
      id: uuid(),
      name: "Dank shoes",
      recipient: "Megan",
    },
  ]);

  const handleReset = () => {
    setCurrentGifts([]);
  };

  const childClick = (newGift) => {
    console.log(newGift);
    setCurrentGifts([...currentGifts, newGift]);
  };

  const removeGift = (giftId) => {
    setCurrentGifts(currentGifts.filter((gift) => gift.id !== giftId));
  };

  return (
    <Fragment>
      <Typography id="test" variant="h5">
        Test
      </Typography>
      <Grid container>
        <Grid item>
          {currentGifts.map((gift) => (
            <Grid container key={gift.id}>
              <Typography
                id="gift"
                variant="body2"
                style={{
                  display: "flex",
                  justifyContent: "center",
                  alignItems: "center",
                  margin: "0.5rem",
                }}
              >
                Gift: {gift.name} for {gift.recipient}
              </Typography>
              <Button
                id="removeGift"
                variant="outlined"
                color="secondary"
                onClick={() => removeGift(gift.id)}
              >
                {"  "}x
              </Button>
            </Grid>
          ))}
          <GiftForm
            currentGifts={currentGifts}
            setCurrentGifts={setCurrentGifts}
            handleChildClick={childClick}
          />
          <Button
            id="resetGifts"
            onClick={handleReset}
            color="default"
            variant="outlined"
            disabled={currentGifts.length > 0 ? false : true}
          >
            Reset
          </Button>
        </Grid>
      </Grid>
    </Fragment>
  );
};

export default App;
