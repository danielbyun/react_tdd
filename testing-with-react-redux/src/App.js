import React from "react";
import Search from "./Tests/SearchFeature/Search";
import CustomEffectHook from "./CustomHooks/CustomEffect/CustomEffectHook";
import CustomStateHook from "./CustomHooks/CustomState/CustomStateHook";
import CustomSubscriptionHook from "./CustomHooks/CustomSubscription/CustomSubscriptionHook";

const App = () => {
  return (
    <div className="App">
      <CustomEffectHook />
      <CustomStateHook />
      <Search />
      <CustomSubscriptionHook />
    </div>
  );
};

export default App;
