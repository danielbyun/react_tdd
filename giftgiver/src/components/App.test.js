/**
 * @jest-environment jsdom
 */

import React from "react";

// helper function from enzyme
// shallowly render a react component
import { configure, mount } from "enzyme";
import { cleanup } from "@testing-library/react";
import Adapter from "enzyme-adapter-react-16";
import App from "./App.js";

configure({ adapter: new Adapter() });

// testing for state inside hooks
const gifts = [
  { id: 1, name: "PS5", recipient: "Mark" },
  { id: 2, name: "Dank shoes", recipient: "Megan" },
];

// unit test
let app = mount(<App gifts={gifts} />);

describe("Gift List App", () => {
  it("renders correctly", () => {
    expect(app).toMatchSnapshot();
  });

  const setState = jest.fn();
  const useStateSpy = jest.spyOn(React, "useState");
  useStateSpy.mockImplementation((init) => [init, setState]);

  beforeEach(() => {
    app = mount(<App gifts={gifts} />);
  });

  afterEach(() => {
    jest.clearAllMocks();
    cleanup();
  });

  // init state
  describe("initial gift state", () => {
    it("initial gift list", () => {
      const giftTexts = app
        .find("#gift")
        .hostNodes()
        .map((node) => node.text());
      expect(giftTexts).toEqual([
        `Gift: ${gifts[0].name} for ${gifts[0].recipient}`,
        `Gift: ${gifts[1].name} for ${gifts[1].recipient}`,
      ]);
    });
  });

  // UI Testing
  describe("ui tests", () => {
    it("remove gift button renders", () => {
      const removeButton = app.find("#removeGift").hostNodes();
      expect(removeButton).toHaveLength(gifts.length);
    });

    it("reset gift button renders", () => {
      const resetButton = app.find("#resetGifts").hostNodes();
      expect(resetButton).not.toBeNull();
    });
  });

  describe("gift actions", () => {
    const pTags = app.find("p#gift").hostNodes();
    const removeButton = app.find("#removeGift").hostNodes();
    // remove a gift
    describe("remove gift", () => {
      it("gifts are rendered", () => {
        // find p tags
        expect(pTags).toHaveLength(2);
      });

      it("remove button is rendered", () => {
        expect(removeButton).toHaveLength(gifts.length);
      });

      it("remove one gift on click", () => {
        // get one button
        removeButton.at(0).simulate("click");

        // test if the state is updated
        expect(setState).toBeTruthy();
      });
    });
    // reset gift
    describe("reset gift", () => {
      it("reset gift list on click", () => {
        app.find("#resetGifts").hostNodes().simulate("click");
        expect(setState).toBeTruthy();
      });
    });
  });
});
