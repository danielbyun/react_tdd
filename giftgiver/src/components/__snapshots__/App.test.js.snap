// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gift List App renders correctly 1`] = `ReactWrapper {}`;
